File: UDF Dönüştürücü/Core/Tools/OCRService.swift
import Foundation
import PDFKit
import UIKit
import Vision

enum OCRError: LocalizedError {
    case cannotOpen
    case noText
    case recognitionFailed(String)

    var errorDescription: String? {
        switch self {
        case .cannotOpen:
            return "Dosya açılamadı."
        case .noText:
            return "Belgede okunabilir metin bulunamadı."
        case .recognitionFailed(let detail):
            return "Metin tanıma hatası: \(detail)"
        }
    }
}

/// Taranmış PDF veya görüntülerden Vision framework ile metin tanır (cihaz üzerinde, sunucusuz).
final class OCRService {

    /// PDF'in tüm sayfalarında OCR çalıştırır; sayfa metinlerini boş satırla ayırarak birleştirir.
    static func recognizeText(pdfURL: URL, progress: @escaping @MainActor (Int, Int) -> Void) async throws -> String {
        guard let document = PDFDocument(url: pdfURL), document.pageCount > 0 else {
            throw OCRError.cannotOpen
        }

        var pageTexts: [String] = []
        for pageIndex in 0..<document.pageCount {
            guard let page = document.page(at: pageIndex) else { continue }
            let bounds = page.bounds(for: .mediaBox)

            // ~300dpi hedefle, çok büyük sayfalarda 4000px ile sınırla.
            let targetScale = min(300.0 / 72.0, 4000.0 / max(bounds.width, bounds.height))
            let renderSize = CGSize(width: bounds.width * targetScale, height: bounds.height * targetScale)
            let bitmap = page.thumbnail(of: renderSize, for: .mediaBox)

            guard let cgImage = bitmap.cgImage else { continue }
            let text = try await recognize(cgImage: cgImage)
            if !text.isEmpty {
                pageTexts.append(text)
            }
            await progress(pageIndex + 1, document.pageCount)
        }

        let fullText = pageTexts.joined(separator: "\n\n")
        guard !fullText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw OCRError.noText
        }
        return fullText
    }

    /// Tek bir görüntü dosyasında OCR çalıştırır.
    static func recognizeText(imageURL: URL) async throws -> String {
        guard let image = UIImage(contentsOfFile: imageURL.path),
              let cgImage = image.cgImage else {
            throw OCRError.cannotOpen
        }
        let text = try await recognize(cgImage: cgImage)
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw OCRError.noText
        }
        return text
    }

    // MARK: - Vision

    private static func recognize(cgImage: CGImage) async throws -> String {
        try await withCheckedThrowingContinuation { continuation in
            let request = VNRecognizeTextRequest { request, error in
                if let error {
                    continuation.resume(throwing: OCRError.recognitionFailed(error.localizedDescription))
                    return
                }
                let observations = (request.results as? [VNRecognizedTextObservation]) ?? []
                let lines = observations.compactMap { $0.topCandidates(1).first?.string }
                continuation.resume(returning: lines.joined(separator: "\n"))
            }
            request.recognitionLevel = .accurate
            request.recognitionLanguages = ["tr-TR", "en-US"]
            request.usesLanguageCorrection = true

            DispatchQueue.global(qos: .userInitiated).async {
                let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
                do {
                    try handler.perform([request])
                } catch {
                    continuation.resume(throwing: OCRError.recognitionFailed(error.localizedDescription))
                }
            }
        }
    }

    // MARK: - Dışa aktarım

    /// OCR metnini mevcut dönüştürme motoruyla DOCX'e yazar.
    static func exportDOCX(text: String, baseName: String) throws -> URL {
        try WordConverter.convert(document: plainDocument(text: text, fileName: baseName))
    }

    /// OCR metnini mevcut dönüştürme motoruyla PDF'e yazar.
    static func exportPDF(text: String, baseName: String) throws -> URL {
        try PDFConverter.convert(document: plainDocument(text: text, fileName: baseName))
    }

    /// OCR metnini UDF olarak yazar.
    static func exportUDF(text: String, baseName: String) throws -> URL {
        let paragraphs = text.components(separatedBy: "\n").map { line in
            UDFCreator.InputParagraph(
                runs: [UDFCreator.InputRun(
                    text: line, isBold: false, isItalic: false, isUnderline: false,
                    fontSize: 12, fontFamily: "Times New Roman"
                )],
                alignment: 0
            )
        }
        return try UDFCreator.create(fileName: "OCR_\(baseName)", paragraphs: paragraphs)
    }

    private static func plainDocument(text: String, fileName: String) -> UDFDocument {
        UDFDocument(
            fileName: "OCR_\(fileName)",
            content: UDFContent(
                text: text,
                rawContent: text,
                contentType: .plainText,
                sections: [UDFSection(title: nil, body: text, level: 0)],
                tables: [],
                isRTF: false,
                formattedString: nil
            ),
            metadata: nil,
            pageFormat: nil
        )
    }
}

File: android/app/src/main/java/com/velikececi/udfdonusturucu/core/tools/OcrService.kt
package com.velikececi.udfdonusturucu.core.tools

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
// play-services-mlkit-text-recognition (unbundled) modülünde TextRecognizerOptions, bundled
// com.google.mlkit:text-recognition'daki gibi com.google.mlkit.vision.text altında değil,
// .latin alt paketindedir.
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.velikececi.udfdonusturucu.core.converters.PdfConverter
import com.velikececi.udfdonusturucu.core.converters.UdfCreator
import com.velikececi.udfdonusturucu.core.converters.WordConverter
import com.velikececi.udfdonusturucu.core.model.ExtractedParagraph
import com.velikececi.udfdonusturucu.core.model.ExtractedTextRun
import com.velikececi.udfdonusturucu.core.model.UdfContent
import com.velikececi.udfdonusturucu.core.model.UdfContentType
import com.velikececi.udfdonusturucu.core.model.UdfDocument
import com.velikececi.udfdonusturucu.core.model.UdfSection
import kotlinx.coroutines.tasks.await
import java.io.File

sealed class OcrException(message: String) : Exception(message) {
    class CannotOpen : OcrException("Dosya açılamadı.")
    class NoText : OcrException("Belgede okunabilir metin bulunamadı.")
    class RecognitionFailed(detail: String?) : OcrException("Metin tanıma başarısız: ${detail ?: "bilinmeyen hata"}")
}

/**
 * OCRService.swift'in Kotlin karşılığı. iOS Vision çerçevesi (`tr-TR`+`en-US`,
 * `usesLanguageCorrection`) kullanıyordu; ML Kit v2'de dil seçeneği/dil-özel düzeltme YOKTUR —
 * `TextRecognizerOptions.DEFAULT_OPTIONS` (Latin script) Türkçe diakritikleri (ç ğ ı İ ö ş ü)
 * kapsar ama iOS'un `tr-TR` sonucunun biraz altında kalabilir.
 *
 * Sayfa rasterizasyonu için [PdfRasterizer] kullanılır (~300 dpi, uzun kenar 4000 px'te
 * sınırlı — iOS'un `min(300/72, 4000/max(w,h))` ölçek hesabıyla aynı).
 */
object OcrService {

    private const val TARGET_DPI = 300f
    private const val MAX_PIXELS = 4000

    private val recognizer by lazy { TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS) }

    /**
     * PDF sayfalarını tek tek tanır. [PdfRasterizer.rasterize]'ın senkron geri çağrısı içinde
     * `Tasks.await(...)` (bloklayan) kullanılır — bu fonksiyon her zaman bir arka plan
     * dispatcher'ından (`Dispatchers.Default`) çağrıldığından ana iş parçacığını asla bloklamaz;
     * askıda (suspend) `.await()`'i senkron bir geri çağrı içine iç içe `runBlocking` ile
     * sokmaktan daha temizdir.
     */
    suspend fun recognizeText(pdfFile: File, context: Context, onProgress: (current: Int, total: Int) -> Unit = { _, _ -> }): String {
        val pageTexts = mutableListOf<String>()
        var failure: Exception? = null

        try {
            PdfRasterizer.rasterize(pdfFile, context, TARGET_DPI, MAX_PIXELS) { index, total, page ->
                if (failure == null) {
                    try {
                        onProgress(index + 1, total)
                        pageTexts += recognizeBitmapBlocking(page.bitmap)
                    } catch (e: Exception) {
                        failure = e
                    }
                }
                page.bitmap.recycle()
            }
        } catch (e: PdfRasterizer.UnsupportedPdfException) {
            throw OcrException.CannotOpen()
        } catch (e: Exception) {
            throw OcrException.CannotOpen()
        }

        failure?.let { throw OcrException.RecognitionFailed(it.message) }

        val combined = pageTexts.filter { it.isNotBlank() }.joinToString("\n\n")
        if (combined.isBlank()) throw OcrException.NoText()
        return combined
    }

    suspend fun recognizeText(imageUri: Uri, context: Context): String {
        val image = try {
            InputImage.fromFilePath(context, imageUri)
        } catch (e: Exception) {
            throw OcrException.CannotOpen()
        }
        val text = try {
            recognizer.process(image).await().text
        } catch (e: Exception) {
            throw OcrException.RecognitionFailed(e.message)
        }
        if (text.isBlank()) throw OcrException.NoText()
        return text
    }

    private fun recognizeBitmapBlocking(bitmap: Bitmap): String {
        val image = InputImage.fromBitmap(bitmap, 0)
        val result = Tasks.await(recognizer.process(image))
        return result.textBlocks
            .sortedWith(compareBy({ it.boundingBox?.top ?: 0 }, { it.boundingBox?.left ?: 0 }))
            .joinToString("\n") { block -> block.lines.joinToString("\n") { it.text } }
    }

    // MARK: - Dışa aktarım

    fun exportDocx(text: String, baseName: String, context: Context): File =
        WordConverter.convert(plainTextDocument(text, baseName), context)

    fun exportPdf(text: String, baseName: String, context: Context): File =
        PdfConverter.convert(plainTextDocument(text, baseName), context)

    fun exportUdf(text: String, baseName: String, context: Context): File {
        val paragraphs = text.split("\n").map { line ->
            ExtractedParagraph(
                runs = listOf(ExtractedTextRun(line, isBold = false, isItalic = false, isUnderline = false, fontSize = 12f, fontFamily = "Times New Roman")),
                alignment = 3,
            )
        }
        return UdfCreator.create("OCR_$baseName", paragraphs, context)
    }

    private fun plainTextDocument(text: String, baseName: String) = UdfDocument(
        fileName = baseName,
        content = UdfContent(
            text = text,
            rawContent = text,
            contentType = UdfContentType.PLAIN_TEXT,
            sections = listOf(UdfSection(title = null, body = text, level = 0)),
            tables = emptyList(),
            isRtf = false,
            paragraphs = emptyList(),
        ),
        metadata = null,
        pageFormat = null,
    )
}
